internal/attestation/aws/snp/validator.go

```diff
@@ -9,99 +9,211 @@ package snp
+func (v *Validator) getTrustedKey(_ context.Context, attDoc vtpm.AttestationDocument, _ []byte) (crypto.PublicKey, error) {
 	pubArea, err := tpm2.DecodePublic(attDoc.Attestation.AkPub)
 	if err != nil {
-		return nil, err
+		return nil, newDecodeError(err)
+	}
+
+	pubKey, err := pubArea.Key()
+	if err != nil {
+		return nil, fmt.Errorf("getting public key: %w", err)
+	}
+
+	akDigest, err := sha512sum(pubKey)
+	if err != nil {
+		return nil, fmt.Errorf("calculating hash of attestation key: %w", err)
+	}
+
+	if err := v.reportValidator.validate(attDoc, v.ask, v.ark, akDigest, v.cfg, v.log); err != nil {
+		return nil, fmt.Errorf("validating SNP report: %w", err)
 	}
 
 	return pubArea.Key()
 }
 
-// tpmEnabled verifies if the virtual machine has the tpm2.0 feature enabled.
-func (v *Validator) tpmEnabled(attestation vtpm.AttestationDocument, _ *attest.MachineState) error {
-	// https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/verify-nitrotpm-support-on-ami.html
-	// 1. Get the vm's ami (from IdentiTyDocument.imageId)
-	// 2. Check the value of key "TpmSupport": {"Value": "v2.0"}"
-	ctx := context.Background()
-
-	idDocument := imds.InstanceIdentityDocument{}
-	err := json.Unmarshal(attestation.InstanceInfo, &idDocument)
+// sha512sum PEM-encodes a public key and calculates the SHA512 hash of the encoded key.
+func sha512sum(key crypto.PublicKey) ([64]byte, error) {
+	pub, err := x509.MarshalPKIXPublicKey(key)
 	if err != nil {
-		return err
+		return [64]byte{}, fmt.Errorf("marshalling public key: %w", err)
+	}
+
+	return sha512.Sum512(pub), nil
+}
+
+// Validate a given SNP report.
+type snpReportValidator interface {
+	validate(attestation vtpm.AttestationDocument, ask *x509.Certificate, ark *x509.Certificate, ak [64]byte, config *config.AWSSEVSNP, log attestation.Logger) error
+}
+
+// Validation logic for the AWS SNP implementation.
+type awsValidator struct {
+	verifier    reportVerifier
+	validator   reportValidator
+	httpsGetter trust.HTTPSGetter
+}
+
+type reportVerifier interface {
+	SnpAttestation(att *sevsnp.Attestation, opts *verify.Options) error
+}
+type reportValidator interface {
+	SnpAttestation(att *sevsnp.Attestation, opts *validate.Options) error
+}
+
+type reportValidatorImpl struct{}
+
+func (r *reportValidatorImpl) SnpAttestation(att *sevsnp.Attestation, opts *validate.Options) error {
+	return validate.SnpAttestation(att, opts)
+}
+
+type reportVerifierImpl struct{}
+
+func (r *reportVerifierImpl) SnpAttestation(att *sevsnp.Attestation, opts *verify.Options) error {
+	return verify.SnpAttestation(att, opts)
+}
+
+// validate the report by checking if it has a valid VLEK signature.
+// The certificate chain ARK -> ASK -> VLEK is also validated.
+// Checks that the report's userData matches the connection's userData.
+func (a *awsValidator) validate(attestation vtpm.AttestationDocument, ask *x509.Certificate, ark *x509.Certificate, akDigest [64]byte, config *config.AWSSEVSNP, log attestation.Logger) error {
```

Comment:
Note: We could possibly make this more generic at some point, as it shares a lot of logic with the Azure validator. But not in this PR.